pkg/virt-handler/device-manager/pci_device.go

```diff
@@ -0,0 +1,428 @@
+	// https://github.com/kubevirt/kubevirt/issues/1196
+	// This can safely be removed once supported upstream Kubernetes is 1.10.3 or higher.
+	emptyList := []*pluginapi.Device{}
+	s.Send(&pluginapi.ListAndWatchResponse{Devices: emptyList})
+
+	s.Send(&pluginapi.ListAndWatchResponse{Devices: dpi.devs})
+
+	for {
+		select {
+		case unhealthy := <-dpi.unhealthy:
+			for _, dev := range dpi.devs {
+				if unhealthy == dev.ID {
+					dev.Health = pluginapi.Unhealthy
+				}
+			}
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: dpi.devs})
+		case healthy := <-dpi.healthy:
+			for _, dev := range dpi.devs {
+				if healthy == dev.ID {
+					dev.Health = pluginapi.Healthy
+				}
+			}
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: dpi.devs})
+		case <-dpi.stop:
+			return nil
+		case <-dpi.done:
+			return nil
+		}
+	}
+}
+
+func formatVFIODeviceSpecs(devID string) []*pluginapi.DeviceSpec {
+	// always add /dev/vfio/vfio device as well
+	devSpecs := make([]*pluginapi.DeviceSpec, 0)
+	devSpecs = append(devSpecs, &pluginapi.DeviceSpec{
+		HostPath:      vfioMount,
+		ContainerPath: vfioMount,
+		Permissions:   "mrw",
+	})
+
+	vfioDevice := filepath.Join(vfioDevicePath, devID)
+	devSpecs = append(devSpecs, &pluginapi.DeviceSpec{
+		HostPath:      vfioDevice,
+		ContainerPath: vfioDevice,
+		Permissions:   "mrw",
+	})
+	return devSpecs
+}
+
+func (dpi *PCIDevicePlugin) Allocate(ctx context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	resourceName := dpi.deviceName
+	resourceNameEnvVar := util.ResourceNameToEnvVar(PCI_RESOURCE_PREFIX, resourceName)
+	allocatedDevices := []string{}
+	resp := new(pluginapi.AllocateResponse)
+	containerResponse := new(pluginapi.ContainerAllocateResponse)
+
+	for _, request := range r.ContainerRequests {
+		deviceSpecs := make([]*pluginapi.DeviceSpec, 0)
+		for _, devID := range request.DevicesIDs {
+			// translate device's iommu group to its pci address
+			devPCIAddress, exist := dpi.iommuToPCIMap[devID]
+			if !exist {
+				continue
+			}
+			allocatedDevices = append(allocatedDevices, devPCIAddress)
+			deviceSpecs = append(deviceSpecs, formatVFIODeviceSpecs(devID)...)
+		}
+		containerResponse.Devices = deviceSpecs
+		envVar := make(map[string]string)
+		envVar[resourceNameEnvVar] = strings.Join(allocatedDevices, ",")
+
+		containerResponse.Envs = envVar
+		resp.ContainerResponses = append(resp.ContainerResponses, containerResponse)
+	}
+	return resp, nil
+}
+
+func (dpi *PCIDevicePlugin) healthCheck() error {
+	logger := log.DefaultLogger()
+	monitoredDevices := make(map[string]string)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to creating a fsnotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	// This way we don't have to mount /dev from the node
+	devicePath := filepath.Join(dpi.deviceRoot, dpi.devicePath)
+
+	// Start watching the files before we check for their existence to avoid races
+	dirName := filepath.Dir(devicePath)
```

Comment:
Does the device plugin need to monitor `/proc/1/root/dev` instead of `/proc/1/root/dev/vfio`?
https://go.dev/play/p/60lVDL3glLb